v-next/hardhat/src/internal/builtin-plugins/network-manager/type-validation.ts

```diff
@@ -221,54 +220,102 @@ const baseNetworkUserConfigSchema = z.discriminatedUnion("type", [
 ]);
 
 function refineEdrNetworkUserConfig(
-  networkConfig: z.infer<typeof baseNetworkUserConfigSchema>,
+  networkConfig: NetworkUserConfig,
   ctx: RefinementCtx,
 ): void {
   if (networkConfig.type === "edr") {
     const {
-      hardfork = LATEST_HARDFORK,
+      chainType = GENERIC_CHAIN_TYPE,
+      hardfork,
+      chains,
       minGasPrice,
       initialBaseFeePerGas,
       enableTransientStorage,
     } = networkConfig;
 
-    if (hardforkGte(hardfork, HardforkName.LONDON)) {
-      if (minGasPrice !== undefined) {
+    if (hardfork !== undefined && !isValidHardforkName(hardfork, chainType)) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        path: ["hardfork"],
+        message: `Invalid hardfork name ${hardfork} for chainType ${chainType}. Expected ${getHardforks(
+          chainType,
+        ).join(" | ")}.`,
+      });
+    }
+
+    if (chains !== undefined) {
+      let chainIdx = 0;
+      for (const [chainId, chainConfig] of chains) {
+        if (chainConfig.hardforkHistory === undefined) {
+          chainIdx++;
+          continue;
+        }
+
+        const chainLevelChainType = chainConfig.chainType ?? GENERIC_CHAIN_TYPE;
+        let hardforkHistoryIdx = 0;
+        for (const [hardforkName] of chainConfig.hardforkHistory) {
+          if (!isValidHardforkName(hardforkName, chainLevelChainType)) {
+            ctx.addIssue({
+              code: z.ZodIssueCode.custom,
+              path: [
+                "chains",
+                chainIdx,
+                "value",
+                "hardforkHistory",
+                hardforkHistoryIdx,
+                "value",
+              ],
+              message: `Invalid hardfork name ${hardforkName} found in chain ${chainId}. Expected ${getHardforks(
+                chainLevelChainType,
+              ).join(" | ")}.`,
+            });
+          }
+          hardforkHistoryIdx++;
+        }
+        chainIdx++;
+      }
+    }
+
+    const resolvedHardfork = hardfork ?? getLatestHardfork(chainType);
+    if (chainType === L1_CHAIN_TYPE || chainType === GENERIC_CHAIN_TYPE) {
```

Comment:
I’ve added an extra validation here, since this should now only be validated for `"l1"` or `"generic"` chain types.

Reply:
Looks correct to me